extensions/servlet/src/main/java/com/stormpath/shiro/servlet/env/StormpathShiroIniEnvironment.java

```diff
@@ -0,0 +1,178 @@
+ *
+ * stormpathRealm.applicationRestUrl = stormpath.application.href
+ * </pre></code>
+ */
+public class StormpathShiroIniEnvironment extends IniWebEnvironment {
+
+    final private Logger log = LoggerFactory.getLogger(StormpathShiroIniEnvironment.class);
+
+    private ConfigLoader configLoader;
+
+    @Override
+    public void setIni(Ini ini) {
+
+        if (ini == null) {
+            ini = new Ini();
+        }
+
+        if (CollectionUtils.isEmpty(ini.getSection(Ini.DEFAULT_SECTION_NAME))) {
+            ini.setSectionProperty(Ini.DEFAULT_SECTION_NAME, "__empty_property__", String.class.getName());
+        }
+
+        super.setIni(ini);
+    }
+
+    private Ini.Section getConfigSection() {
+        Ini ini = getIni();
+
+        Ini.Section configSection = ini.getSection(IniSecurityManagerFactory.MAIN_SECTION_NAME);
+        if (CollectionUtils.isEmpty(configSection)) {
+            configSection = ini.getSection(Ini.DEFAULT_SECTION_NAME);
+        }
+
+        return configSection;
+    }
+
+    private Map<String, ?> getDefaultEnvironmentObjects() {
+
+        Map<String, Object> defaults = new LinkedHashMap<String, Object>();
+
+        Ini.Section configSection = getConfigSection();
+        defaults.put("stormpathClient", new StormpathWebClientFactory(getServletContext()));
+
+        String href = ConfigResolver.INSTANCE.getConfig(getServletContext()).get(DefaultServletContextClientFactory.STORMPATH_APPLICATION_HREF);
+
+        ApplicationRealm stormpathRealm = new StormpathWebRealm();
+        stormpathRealm.setApplicationRestUrl(href);
+        defaults.put("stormpathRealm", stormpathRealm);
+
+        // lazy associate the client with the realm, so changes can be made if needed.
+        if (!configSection.containsKey("stormpathRealm.client")) {
+            configSection.put("stormpathRealm.client", "$stormpathClient");
+        }
+
+        return defaults;
+    }
+
+
+    @Override
+    protected void configure() {
+
+        configureStormpathEnvironment();
+
+        super.configure();
+
+        ClientFactory clientFactory = getObject("stormpathClient", ClientFactory.class);
+        log.debug("Updating Client in ServletContext, with instance configured via shiro.ini");
+        getServletContext().setAttribute(Client.class.getName(), clientFactory.getInstance());
+
+    }
+
+    protected void configureStormpathEnvironment() {
+        ensureConfigLoader().createConfig(getServletContext());
+    }
+
+    private ConfigLoader ensureConfigLoader() {
+        if (configLoader == null) {
+            configLoader = new ShiroIniConfigLoader(getIni());
+        }
+        return configLoader;
+    }
+
+    @Override
+    public void destroy() throws Exception {
+        ensureConfigLoader().destroyConfig(getServletContext());
+        getServletContext().removeAttribute(Client.class.getName());
+        super.destroy();
+    }
+
+    // TODO think about putting these in Shiro proper
```

Comment:
(this would require a shiro 1.4 release, so we should probably do both, leave this code here until it can be backported and released.